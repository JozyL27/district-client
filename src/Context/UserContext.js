import React, { Component } from "react";
import TokenService from "../services/token-service";
import io from "socket.io-client";
import config from "../config";

const UserContext = React.createContext({
  user: {},
  error: null,
  socket: null,
  setError: () => {},
  clearError: () => {},
  setUser: () => {},
  processLogin: () => {},
  processLogout: () => {},
});

export default UserContext;

export class UserProvider extends Component {
  constructor(props) {
    super(props);
    const state = { user: {}, error: null, socket: null };

    const jwtPayload = TokenService.parseAuthToken();

    if (jwtPayload) {
      const socket = io(config.SOCKET_CONNECTION);
      state.user = {
        id: jwtPayload.user_id,
        email: jwtPayload.email,
        username: jwtPayload.sub,
      };
      state.socket = socket.connect();
    }

    this.state = state;
  }

  setError = (error) => {
    this.setState({ error });
  };

  clearError = () => {
    this.setState({ error: null });
  };

  setUser = (user) => {
    this.setState({ user });
  };

  processLogin = (authToken) => {
    TokenService.saveAuthToken(authToken);
    const jwtPayload = TokenService.parseAuthToken();
    const socket = io(config.SOCKET_CONNECTION);
    this.setUser({
      id: jwtPayload.user_id,
      email: jwtPayload.email,
      username: jwtPayload.sub,
    });
    this.setState({ socket: socket.connect() });
  };

  processLogout = () => {
    TokenService.clearAuthToken();
    this.setUser({});
    this.state.socket.disconnect();
    this.setState({ socket: null });
  };

  render() {
    const values = {
      user: this.state.user,
      error: this.state.error,
      socket: this.state.socket,
      setError: this.setError,
      clearError: this.clearError,
      setUser: this.setUser,
      processLogin: this.processLogin,
      processLogout: this.processLogout,
    };
    return (
      <UserContext.Provider value={values}>
        {this.props.children}
      </UserContext.Provider>
    );
  }
}
