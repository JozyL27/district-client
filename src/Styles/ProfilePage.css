.profilePageContainer {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-top: 50px;
    margin-bottom: 50px;
}

.userInfoContainer {
    display: flex;
    flex-direction: column;
    align-items: center;
    max-width: 700px;
}

.profileAvatar {
    max-height: 200px;
    margin-bottom: 25px;
    border-radius: 50%;
    background: #ffffff;
    box-shadow:  -32px 32px 64px #bfbfbf, 
                32px -32px 64px #ffffff;
}

.profileButtonContainer {
    margin: 20px auto 20px auto;
}

.profileFabContainer {
    margin: 20px auto 20px auto;
}

.profileArticles {
    margin: 0;
    padding: 0;
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    justify-content: space-evenly;
    margin-bottom: 30px;
}

.bioContainer {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 10px;
}

.profileBio {
    text-align: center;
    line-height: 1.4;
    max-width: 90%;
}

.profileUsername {
    padding-top: 20px;
    font-weight: 500;
    text-align: center;
    font-size: 20px;
}

@media screen and (min-width: 768px) {
    .userInfoContainer {
        flex-direction: row;
        margin-bottom: 50px;
    }

    .profilePageContainer {
        margin-top: 100px;
    }

    .profileAvatar {
        margin-bottom: 0px;
    }
}