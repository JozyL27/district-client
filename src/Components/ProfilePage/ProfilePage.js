import React, { Component } from 'react'
import UserContext from '../../Context/UserContext'
import UserService from '../../services/user-service'
import ArticlesService from '../../services/article-service'
import Fab from '@material-ui/core/Fab'
import AddIcon from '@material-ui/icons/Add'
import Button from '@material-ui/core/Button'
import ArticleCard from '../ArticleCard/ArticleCard'
import NavArrows from '../NavArrows/NavArrows'
import '../../Styles/ProfilePage.css'
import avatar from '../../illustrations/01.png'
import EditProfileCard from '../EditProfileCard/EditProfileCard'


export default class ProfilePage extends Component {
    static contextType = UserContext
    state = { articles: [], userInfo: {}, 
    error: null, page: 1, isEditing: false, bio: '', username: '' }

    componentDidMount() {
        const { user } = this.context || {}
        const { page } = this.state
        UserService.getAuthorInfo(user.id)
        .then(res => {
            res.error ? this.setState({ error: res.error })
            : this.setState({ userInfo: res, bio: res.bio, 
                username: res.username })

            ArticlesService.getMyArticles(user.id, page)
            .then(res => res.error ?
                this.setState({ error: res.error})
                : this.setState({ articles: res })
                )
        })
    }

    handleEditButton = () => {
        let { isEditing } = this.state
        this.setState({ isEditing: !isEditing })
    }

    handleCancelButton = () => {
        let { isEditing, userInfo } = this.state
        this.setState({ isEditing: !isEditing, bio: userInfo.bio })
    }

    handleChange = (event) => {
        this.setState({
            [event.target.name]: event.target.value
        })
    }

    render() {
        const { userInfo, articles, page, 
            isEditing, bio, username } = this.state || {}
            console.log(bio, username)
        return (
            <section className='profilePageContainer'>
                {!isEditing ?
                <>
                    <div className='userInfoContainer'>
                        <img 
                        src={avatar} 
                        alt='avatar' 
                        className='profileAvatar'
                        />
                        <div className='bioContainer'>
                            <span className='profileUsername'>{userInfo.username}</span>
                            {userInfo.bio && <p className='profileBio'>{userInfo.bio}</p>}
                        </div>
                    </div>
                    <div className='profileButtonContainer'>
                        <Button 
                        variant='contained' 
                        color='primary'
                        onClick={this.handleEditButton}
                        >
                            Edit Profile
                        </Button>
                    </div>
                </>
                : 
                <EditProfileCard 
                avatar={avatar}
                username={username}
                bio={bio}
                handleBioChange={this.handleChange}
                handleCancelButton={this.handleCancelButton}
                />
                }
                <div className='profileFabContainer'>
                    <Fab>
                        <AddIcon />
                    </Fab>
                </div>
                <ul className='profileArticles'>
                    {articles.length > 0 && articles.map(article => {
                        return (
                        <ArticleCard 
                        id={article.id}
                        key={article.id}
                        username={userInfo.username}
                        title={article.title}
                        upvotes={article.upvotes}
                        date_published={article.date_published}
                        />
                        )
                    })}
                </ul>
                {articles.length === 9 &&
                <NavArrows 
                styleName='profileNavArrows'
                page={page}
                />}
            </section>
        )
    }
}