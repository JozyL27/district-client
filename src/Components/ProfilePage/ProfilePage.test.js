import React from "react";
import ReactDom from "react-dom";
import renderer from "react-test-renderer";
import ProfilePage from "./ProfilePage";
import { UserProvider } from "../../Context/UserContext";
import { MemoryRouter } from "react-router-dom";

describe("<ProfilePage />", () => {
  // smoke test
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDom.render(
      <MemoryRouter>
        <UserProvider>
          <ProfilePage />
        </UserProvider>
      </MemoryRouter>,
      div
    );
    ReactDom.unmountComponentAtNode(div);
  });

  // snapshot
  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <UserProvider>
            <ProfilePage />
          </UserProvider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
