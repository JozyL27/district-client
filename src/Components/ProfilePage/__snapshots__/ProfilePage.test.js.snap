// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfilePage /> renders the UI as expected 1`] = `
<section
  className="profilePageContainer"
>
  <div
    className="userInfoContainer"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle makeStyles-large-11 MuiAvatar-colorDefault"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiAvatar-fallback"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
    <div
      className="bioContainer"
    >
      <span
        className="profileUsername"
      />
    </div>
  </div>
  <div
    className="followersContainer"
  />
  <div
    className="profileButtonContainer"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Edit Profile
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="profileFabContainer"
  >
    <button
      className="MuiButtonBase-root MuiFab-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiTabs-root"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": null,
            "overflow": "hidden",
          }
        }
      >
        <div
          className="MuiTabs-flexContainer MuiTabs-centered"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              My Articles
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Upvoted Articles
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          className="PrivateTabIndicator-root-17 PrivateTabIndicator-colorPrimary-18 MuiTabs-indicator"
          style={
            Object {
              "left": 0,
              "width": 0,
            }
          }
        />
      </div>
    </div>
  </div>
  <ul
    className="profileArticles"
  />
</section>
`;
